# Vectors

Our programs have all been 2 dimensional, so we need to work with both x and y
values. Now that we have a good understanding of how objects work, we can start
working with vectors, which are objects that combine the x and y values into a
single object. 

There are two ways of thinking about vectors. 
* One is as a position, which is a point in space, which has a X and Y position. 
* The other is as a direction and a magnitude, which is a line with a length and
a direction.

Here is what the vector (x=8, y=8) looks like:

![Vector](images/v88.png)

This vector represents the point (8,8) on a grid, and if we draw a line from the
origin (0,0) to the point (8,8), we get a line that is 11.3 units long, at an
angle of 45 degrees. So, it can both be a point and a line, as long as we can
assume a starting point for the line. 

Here is a really simple Vector class that we can use to represent vectors:



In [6]:

import math

class Vector:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    @property
    def magnitude(self): # magnitude is the length of the vector
        # Magnitude of the vector: sqrt(x^2 + y^2)
        return math.sqrt(self.x**2 + self.y**2)

    @property
    def direction(self):
        # Direction (angle) in radians: atan2(y, x)
        return math.atan2(self.y, self.x) * 180 / math.pi

    def __repr__(self):
        return f"Vector({self.x}, {self.y})"

# Example usage:
v = Vector(4, 4)
print(f"Vector: {v}")
print(f"Magnitude: {v.magnitude}")
print(f"Direction: {v.direction}°")


Vector: Vector(4, 4)
Magnitude: 5.656854249492381
Direction: 45.0°




As you can see, your create the vector by passing in the x and y values, and
then you can calculate the length and angle using trigonometry.

If we have a vector that is a position, we can add two vectors together to get a
new vector that is the sum of the two. This is called vector addition. If we
have a vector that is a direction and magnitude, we can multiply the vector by a
scalar ( a single number ) to get a new vector that is the original vector
multiplied by the scalar. This is called scalar multiplication.

So, suppose that we have a vector called `p1`, which is the position of the player, at
x=100 and y=100. We can write this as:



In [7]:

from pygame.math import Vector2 # Using PyGame's Vector, which has a lot more features

p1 = Vector2(100, 100)

p1

<Vector2(100, 100)>


Now, suppose that we have a vector called `move`, which is the amount we want to
move the player. We can move the player just by adding the move vector to the
player vector. We can write this as:



In [8]:

move = Vector2(10, 0) # Move 10 pixels to the right, none up/down

p1 += move
p1

<Vector2(110, 100)>


Now, `p1` is at x=110 and y=100. To get that, we added the x values of the two
vectors together, and the y values of the two vectors together. So the new x
value is 100 + 10 = 110, and the new y value is 100 + 0 = 100.



For a more detaied example. run the program `01a_vector_example.py.` This program has a custom version
of the vector class that displays on a larger grid. Here is what the programs output looks like:

![Vector Example](images/vector_grid.png)


Here are the vectors that are displayed in the program:

```python 
v0 = Vector20(0,0)
v1 = Vector20(8, 8)  
v2 = Vector20(3, -12)  
v3 = Vector20(-4, -2)  
v4 = Vector20(-12, 0) 
v5 = Vector20(0, 12)
```

Compare the code to the output to see how the vectors are displayed on the grid. Notice that we can move from one place
to another on the grid by adding the vectors together. For example, to move from v0 to v1, we can add v0 to v1. 



## Rotations
Or, suppose that we want to move the player 100 pixels in the direction of 45 degrees. We can write this as:


In [1]:
p1 = Vector2(100, 100)
move = Vector2(100, 0)
move.rotate_ip(45) # rotate the vector "in-place"

p1 += move

p1

NameError: name 'Vector2' is not defined

When you add vectors, you really just add the x and y values together. 


In [2]:
from  pygame import Vector2

v1 = Vector2(10, 10 )
v2 = Vector2(20, 20 )


print(v1 + v2)

[30, 30]


Rotating a vector changes both the x and y values. The x value is changed by 
`cos(angle) * x`, and the y value is changed by `sin(angle) * y`. 


In [4]:
v1 = Vector2(10,10)
move = Vector2(10, 0).rotate(45)

print(v1)
print(move)

print(v1 + move)


[10, 10]
[7.07107, 7.07107]
[17.0711, 17.0711]


## Assignment 1

Draw a square with Vectors. 

1. Open and read `01a_vector_example.py` and run the program.
2. Create vector v1 with x = 0 and y = 1 
3. Scale the vector by 10 with multiplication, then rotate it by 90 degrees so
   it points to the right.
4. Draw the vector from the position (5,5) on the grid.
5. Create a new vector v2 by rotating v1 by 90 dregrees, and draw it from the
   ending position of v1 ( which is the return value from the draw_v20()
   function )
6. Continue rotating and drawing until you have drawn a square. 

Using vectors instead of seperate x and y makes our programs easier to write, read and debug. 


## Next Steps

Your next lesson is [02_Using_Vectors](02_Using_Vectors.ipynb).